core/src/main/scala/kafka/server/metadata/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,200 @@
+  def process(registerLocalBrokerEvent: OutOfBandRegisterLocalBrokerEvent): Unit
+  def process(fenceLocalBrokerEvent: OutOfBandFenceLocalBrokerEvent): Unit
+}
+
+class BrokerMetadataEventManager(config: KafkaConfig,
+                                 time: Time,
+                                 processor: MetadataEventProcessor,
+                                 eventQueueTimeHistogramTimeoutMs: Long = 300000) extends KafkaMetricsGroup {
+
+  private val blockingQueue = new LinkedBlockingQueue[QueuedEvent]
+  private val bufferQueue: util.Queue[QueuedEvent] = new util.ArrayDeque[QueuedEvent]
+  private val outOfBandQueue = new ConcurrentLinkedQueue[QueuedEvent]
+  @volatile private var totalQueueSize: Int = 0
+
+  private val thread = new BrokerMetadataEventThread(
+    s"${BrokerMetadataEventManager.BrokerMetadataEventThreadNamePrefix}${config.brokerId}${BrokerMetadataEventManager.BrokerMetadataEventThreadNameSuffix}")
+
+  // metrics
+  private val eventQueueTimeHist = newHistogram(BrokerMetadataEventManager.EventQueueTimeMetricName)
+  newGauge(BrokerMetadataEventManager.EventQueueSizeMetricName, () => totalQueueSize)
+
+  def start(): Unit = {
+    put(StartupEvent)
+    thread.start()
+  }
+
+  def close(): Unit = {
+    try {
+      thread.initiateShutdown()
+      put(WakeupEvent) // wake up the thread in case it is blocked on queue.take()
+      thread.awaitShutdown()
+    } finally {
+      removeMetric(BrokerMetadataEventManager.EventQueueTimeMetricName)
+      removeMetric(BrokerMetadataEventManager.EventQueueSizeMetricName)
+    }
+  }
+
+  def put(event: BrokerMetadataEvent): QueuedEvent = {
+    val queuedEvent = new QueuedEvent(event, time.milliseconds())
+    event match {
+      case _: OutOfBandRegisterLocalBrokerEvent |
+           _: OutOfBandFenceLocalBrokerEvent =>
+        outOfBandQueue.add(queuedEvent)
+        blockingQueue.add(new QueuedEvent(WakeupEvent, queuedEvent.enqueueTimeMs))
+      case _ => blockingQueue.put(queuedEvent)
+    }
+    queuedEvent
+  }
+
+  def currentMetadataOffset(): Long = processor.currentMetadataOffset()
+
+  private class BrokerMetadataEventThread(name: String) extends ShutdownableThread(name = name, isInterruptible = false) {
+    logIdent = s"[BrokerMetadataEventThread] "
+
+    override def doWork(): Unit = {
+      val dequeued: QueuedEvent = pollFromEventQueue()
+      dequeued.event match {
+        case StartupEvent => processor.processStartup()
+        case metadataEvent: MetadataEvent =>
+          eventQueueTimeHist.update(time.milliseconds() - dequeued.enqueueTimeMs)
+          try {
+            processor.process(metadataEvent.apiMessages, metadataEvent.lastMetadataOffset)
+          } catch {
+            case e: Throwable => error(s"Uncaught error processing event: $metadataEvent", e)
+          }
+        case WakeupEvent => // Ignore since it serves solely to wake us up
+        case registerLocalBrokerEvent: OutOfBandRegisterLocalBrokerEvent =>
```

Comment:
It's clear that we need to handle fencing out-of-band-- or at least, fencing that is caused by slow metadata processing must be handled that way.  It's less clear that we need to handle registration out-of-band.  It seems better to just wait for our own registration record to go by as we follow the metadata log.  Otherwise we have to worry about divergent state, which is the source of a lot of our frustration.

Reply:
I think it might be best to leave this as it is for now and adjust/remove it if necessary as we better understand the implications -- and our ultimate implementation -- of fencing.  The broker may or may not need to know what its broker epoch is.  If the broker ultimately doesn't need to know it's broker epoch then we can get rid of this, but for now I would prefer to keep it under the assumption that the broker will need to know its broker epoch.

Reply:
as per our discussion yesterday, I'm OK with handling unfencing out-of-band.